feat: Queue extra packets in the adapter

simon-lemay-unity opened the pull request:
This addresses MTT-1887.

UTP is limited to 32 in-flight packets on reliable pipelines. When that limit is reached, the error about having "too many in-flight packets" is produced. Many users have reported seeing this error and it's a frustrating one because there's nothing they can do about it (except reducing the scope of their game, which is probably not acceptable).

Increasing the limit in UTP would impose additional overhead to reliable packets, something that the DOTS NetCode team feels strongly against. Increasing the limit without adding this extra overhead would pose a significant engineering challenge. Instead, the chosen solution is to queue packets in the adapter when UTP's limit is reached. This is a similar situation to what DOTS NetCode is doing in a similar situation.

A new configuration setting ('Max Send Queue Size') is introduced to control the size of that queue. The existing 'Send Queue Batch Size' setting remains and controls the maximum size of a single batch of sends (which is effectively the maximum payload size). The naming is a bit unfortunate, but it preserves the previous behavior.

A note on the implementation: the new queue (`BatchedSendQueue`) is written in such a way that it could be passed to a Burst-compiled job. My intention for a future change is to put the batched sending of packets into its own job. This is also why all that code has been moved into the `Update` method.

Once/If approved, I intend to backport these changes to 1.0.0 and 1.1.0 since it resolves an issue that hits many users (the "too many in-flight packets" error).

## Changelog

### com.unity.netcode.adapter.utp

* Added: Added new 'Max Send Queue Size' configuration field in the inspector. This controls the size of the send queue that is used to accumulate small sends together and also acts as an overflow queue when there are too many in-flight packets or when other internal queues are full.
* Fixed: Avoid "too many inflight packets" errors by queueing packets in a queue when the limit of inflight packets is reached in UTP. The size of this queue can be controlled with the 'Max Send Queue Size' configuration field.

## Testing and Documentation

* Includes unit tests.
* Includes integration tests.
* No documentation changes or additions were necessary.

simon-lemay-unity committed:
Add batched send/receive queues

The send queue is meant to be used to store individual messages and read
them back in batched chunks (to be sent on the network). It's also meant
to act as a sort of overflow storage when UTP internal queues are full.

The receive queue is meant to store batched messages (batched by the
send queue) and read them out one at a time.

simon-lemay-unity committed:
Allow tests to access internals

simon-lemay-unity committed:
Remove dummy tests

simon-lemay-unity committed:
Update UnityTransport to use new batched queues

simon-lemay-unity committed:
Update transport tests

simon-lemay-unity committed:
Add latest changes to CHANGELOG

simon-lemay-unity committed:
Minor stylistic improvements for consistency

simon-lemay-unity committed:
Fix whitespace issue

simon-lemay-unity committed:
Ignore test until fix for UTP is released

simon-lemay-unity committed:
Correctly handle ArraySegment not at beginning of array

When handling incoming messages from the transport

Rosme commented on com.unity.netcode.gameobjects/Runtime/Messaging/MessagingSystem.cs:
```diff
@@ -150,7 +150,7 @@ internal void HandleIncomingData(ulong clientId, ArraySegment<byte> data, float
                 {
                     var batchReader =
-                        new FastBufferReader(nativeData, Allocator.None, data.Count, data.Offset);
+                        new FastBufferReader(nativeData + data.Offset, Allocator.None, data.Count);
```
Might have missed something, but why are we moving the `data.Offset` out of the offset argument?

simon-lemay-unity replied:
Yeah I should have added a comment about that change. Basically the previous behavior was wrong when the `ArraySegment` returned by the transport is not for a segment starting at the beginning of its backing array (`data.Offset` greater than 0).

`FastBufferReader` is written such that the length you pass as the third argument here is the full length of the buffer pointed to by the first argument. But here that's not what we're providing; `data.Count` is the length of the _segment_ of the array we're handling and we were passing a pointer to the backing array of that segment as the buffer.

It might be tempting instead to keep the offset and pass the length of the backing array (`data.Array.Length`) instead of `data.Count`, but that's also wrong since the backing array could go on after the end of the segment. The correct value to pass as the length in this case would be `data.Offset + data.Count` but that's kinda confusing. It's simpler to just make it look to `FastBufferReader` like it's dealing with a clean buffer without any starting offset.

ShadauxCat replied:
Is this a change that should be applied to FastBufferReader itself? When I was writing that part of FastBufferReader I debated with myself on whether the third argument should be the total buffer size or the size to use past the offset. I'm willing to accept I may have chosen wrong here.

simon-lemay-unity replied:
> This is an immediate point of concern because it means we'll be fixing this for the people who are currently having this issue by introducing new issues to other users. I don't think we should change the meaning of an existing setting in a way that makes existing configurations invalid. IMO, we should deprecate this setting and make a new one with a sensible default so that users who upgrade don't find themselves suddenly broken.

Fair point. Will deprecate that setting and introduce a new one.

simon-lemay-unity replied:
I don't have any strong opinion on whether that length parameter should be the total length of the buffer or the length past the offset. My concern is more that the documentation of the constructors seems to indicate it's the latter, while internally the code appears to use the former definition.

For the current change, I went with the simplest and smallest targeted fix I could find, considering I'm not familiar with the `FastBufferReader` code and the implications of modifying it.

ShadauxCat commented:
> 'Send Queue Batch Size' now controls the size of that queue instead of the size of a single batch of messages. It should thus be set much higher than it was before.

This is an immediate point of concern because it means we'll be fixing this for the people who are currently having this issue by introducing new issues to other users. I don't think we should change the meaning of an existing setting in a way that makes existing configurations invalid. IMO, we should deprecate this setting and make a new one with a sensible default so that users who upgrade don't find themselves suddenly broken,

simon-lemay-unity commented:
> This is an immediate point of concern because it means we'll be fixing this for the people who are currently having this issue by introducing new issues to other users. I don't think we should change the meaning of an existing setting in a way that makes existing configurations invalid. IMO, we should deprecate this setting and make a new one with a sensible default so that users who upgrade don't find themselves suddenly broken.

Fair point. Will deprecate that setting and introduce a new one.

simon-lemay-unity committed:
Don't change how existing settings behave

m_SendQueueBatchSize was effectively the maximum payload size and this
is what it will stay as. The new maximum payload size setting has been
removed.

The size of the send queue is now controlled by m_MaxSendQueueSize.

simon-lemay-unity commented:
Following comment from @ShadauxCat, changed the behavior to avoid breaking existing user configurations. Instead of reusing the 'Send Queue Batch Size' setting to control the size of the send queue, a new one is introduced ('Max Send Queue Size').

The setting 'Send Queue Batch Size' stays on as the maximum size of a single batch of sends, which is also effectively the maximum payload size (I dropped the new setting for this). This is exactly what that setting meant before (even though the naming is a tad unfortunate).

The description of the PR has been edited accordingly.

simon-lemay-unity committed:
Improve "too many in-flight" error message

simon-lemay-unity committed:
Merge branch 'develop' into feature/larger-adapter-send-queue

simon-lemay-unity committed:
Had left conflict markers in .meta files...

simon-lemay-unity committed:
Merge branch 'develop' into feature/larger-adapter-send-queue

simon-lemay-unity committed:
Stylistic improvements

andrews-unity committed:
Merge branch 'develop' into feature/larger-adapter-send-queue

andrews-unity committed:
Merge branch 'develop' into feature/larger-adapter-send-queue

simon-lemay-unity committed:
Fix standards issues

netcode-ci-service committed:
Merge develop into feature/larger-adapter-send-queue

netcode-ci-service committed:
Merge develop into feature/larger-adapter-send-queue

simon-lemay-unity committed:
Enable test that waited on a fix in UTP 1.0.0-pre.10
